text/0230-simplify-qunit-testing-api.md

```diff
@@ -0,0 +1,466 @@
+- Start Date: 2017-06-13
+- RFC PR: [emberjs/rfcs#232](https://github.com/emberjs/rfcs/pull/232)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+In order to embrace newer features being added by QUnit (our chosen default
+testing framework), we need to reduce the brittle coupling between `ember-qunit`
+and QUnit itself. 
+
+This RFC proposes a new testing syntax, that will expose QUnit API directly while
+also making tests much easier to understand. 
+
+# Motivation
+
+QUnit feature development has been accelerating since the ramp up to QUnit 2.0.
+A number of new features have been added that make testing our applications
+much easier, but the current structure of `ember-qunit` impedes our ability 
+to take advantage of some of these features.
+
+Developers are often confused by our `moduleFor*` APIs, questions like these
+are very common: 
+
+* What "magic" is `ember-qunit` doing? 
+* Where are the lines between QUnit and ember-qunit? 
+* How can I use QUnit for plain JS objects? 
+
+The way that `ember-qunit` wraps QUnit functionality makes the division
+of responsiblity much harder to understand, and leads folks to believe that there
+is much more going on in `ember-qunit` than there is. It should be much clearer
+what `ember-qunit` is responsible for and what we rely on QUnit for.
+
+This RFC also aims to remove a number of custom testing only APIs that exist today
+(largely because the container/registry system was completely private when the
+current tools were authored). Instead of things like `this.subject`, `this.register`,
+`this.inject`, or `this.lookup` we can rely on the standard way of performing these
+functions in Ember via the owner API.
+
+When this RFC has been implemented and rolled out, these questions should all be
+addressed and our testing system will both: embrace QUnit much more **and** 
+be much more framework agnostic, all the while dropping custom testing only APIs
+in favor of public APIs that work across tests and app code.
+
+Sounds like a neat trick, huh?
+
+# Detailed design
+
+The primary change being proposed in this RFC is to migrate to using the QUnit
+nested module syntax, and update our custom setup/teardown into a more functional
+API.
+
+Lets look at a basic example:
+
+```js
+// **** before ****
+
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+moduleForComponent('x-foo', {
+  integration: true
+});
+
+test('renders', function(assert) {
+  assert.expect(1);
+
+  this.render(hbs`{{pretty-color name="red"}}`);
+
+  assert.equal(this.$('.color-name').text(), 'red');
+});
+
+// **** after ****
+
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+module('x-foo', function(hooks) {
+  setupRenderingTest(hooks);
+
+  test('renders', async function(assert) {
+    assert.expect(1);
+
+    await this.render(hbs`{{pretty-color name="red"}}`);
```

Comment:
@rwjblue apologies for wandering by, but seems like continuing `setupRenderingTest`-injects-`render`-into-`this` is continuing the "change objects at runtime" style of ember's JS roots.

If a goal (of some subset of the users/disclaimer IANAE) is to have better TS support, it seems like these sort of "suddenly show up on `this`" methods will be inherently untypeable, and instead `render` should be a the static import (like `test`, `module`, etc.) or else called against some non-`this` variable that can be appropriately typed, e.g. returned from the `setupRenderingTest` method/something like that.

Granted, no idea what compromises you have to make for backwards compatibility, e.g. if having users move from `this.render` to `render` or `foo.render` is just a no-go, but if you're making a big change, seems like moving towards typeable-if-you-want APIs/DSLs would be worthwhile.

Reply:
Subsequent to this RFC I implemented the helpers as importables (in fact that is what the codemod actually does), there is more written on this in #268.

Reply:
Ah great, thanks! I'd read #268 and noticed it was that way there, but didn't put 2+2 together to realize that meant these were older examples that were supplanted by #268's.